src/io/rsm/rsm_client.hpp

```diff
@@ -43,28 +45,48 @@ class RsmClient {
   Address leader_;
   ServerPool server_addrs_;
 
+  /// State for single async read/write operations. In the future this could become a map
+  /// of async operations that can be accessed via an ID etc...
+  std::optional<Time> async_read_before_;
+  std::optional<ResponseFuture<ReadResponse<ReadResponseT>>> async_read_;
+  ReadRequestT current_read_request_;
+
+  std::optional<Time> async_write_before_;
+  std::optional<ResponseFuture<WriteResponse<WriteResponseT>>> async_write_;
+  WriteRequestT current_write_request_;
```

Comment:
I leave this comment for completeness, that is, I don't recommend any action for now. I am concerned for this `State` because once we use a map to allow concurrency we implicitly opt in to a very *synchronized* shared state. What I mean by that, is that the QE will be sending a lot of requests. Each associated `Future` of a request `X` must be removed from that shared state once its fulfilled (same when we insert a future, that is when we make a request). Therefore, in a concurrency setup, we will need to protect this Shared State by a mutex. However, this could be problematic in practice because this data structure is almost guaranteed to become *hot*, and when that happens we will have a lot "waiting time" between threads. That's something we should consider.